src/GafferCycles/IECoreCyclesPreview/Renderer.cpp

```diff
@@ -1752,6 +1752,7 @@ class LightCache : public IECore::RefCounted
 		{
 			ccl::Light *light = new ccl::Light();
 			light->name = nodeName.c_str();
+			light->set_owner( m_scene );
```

Comment:
Since we don't have test coverage for this, I'd like a fuller understanding of why we need it. From reading the Cycles source, I _think_ this is because in `clearUnused()` we call `m_scene->delete_nodes()`. And `Scene::delete_nodes()` asserts that the owner of the node is the scene (internally in `remove_nodes_in_set()`). So what we are fixing is an assertion failure when building against a debug build of Cycles. Is that correct? If so, it would be great to transfer that information into the commit log for future reference.

I assume an alternative fix would be to pass `owner = nullptr` to `m_scene->delete_nodes()`? Do you know anything about the intention behind `Node::owner`? As far as I can tell from the Cycles source, it's not really used for anything other than assertions. And there's also `Node::ref_count` which seems to imply support for shared ownership, but then doesn't really seem to do much either. I'd love to understand the intention behind these two...

Reply:
> Since we don't have test coverage for this, I'd like a fuller understanding of why we need it. From reading the Cycles source, I _think_ this is because in `clearUnused()` we call `m_scene->delete_nodes()`. And `Scene::delete_nodes()` asserts that the owner of the node is the scene (internally in `remove_nodes_in_set()`). So what we are fixing is an assertion failure when building against a debug build of Cycles. Is that correct? If so, it would be great to transfer that information into the commit log for future reference.
> 
Yep asserts, I did this call for other objects like shaders and objects/meshes but forgot the lights. Alright will do.

> I assume an alternative fix would be to pass `owner = nullptr` to `m_scene->delete_nodes()`? Do you know anything about the intention behind `Node::owner`? As far as I can tell from the Cycles source, it's not really used for anything other than assertions. And there's also `Node::ref_count` which seems to imply support for shared ownership, but then doesn't really seem to do much either. I'd love to understand the intention behind these two...

I believe the only real use-case that this is important outside of assets is for the `procedural.h` interface that the Alembic procedural uses to tell cycles that it owns them and will leave it to delete those and not scene. I have thought about doing the same with IECoreCycles being the real owner and probably have a git stash somewhere with a basic sketch but I don't remember now why I abandoned it...